scipy/optimize/_cobyqa_py.py

```diff
@@ -64,4 +68,5 @@ def _minimize_cobyqa(fun, x0, args=(), bounds=None, constraints=(),
         'radius_final': float(final_tr_radius),
         'scale': bool(scale),
     }
-    return minimize(fun, x0, args, bounds, constraints, callback, options)
+    with COBYQA_LOCK:
```

Comment:
@ragonneau this is perhaps of interest to you? I had a quick look through the `cobyqa` package, but it was not easy to tell from a quick browse what wasn't thread-safe.

For context, without this lock, the tests for `cobyqa` within SciPy show that numerically incorrect solutions are returned from `cobyqa` when calling it from multiple threads in parallel (as done by `pytest-run-parallel` in this PR).

Reply:
Hi @rgommers,

I am currently extremely busy, but I will have a look in the future. Do you mean that the module is not deterministic (in the sense that two different solutions are returned when submitting the exact same inputs) when run in parallel? I actually tested this a lot in the past, but I might have missed something...

Cheers,
Tom.

Reply:
Yes indeed, that seems to be the case. `cobyqa` is pure Python, so it's not impossible that it's a bug in CPython's free-threading support (it is specific to free-threading, so that hasn't really been tested before).

Anyway, this is not urgent at all, so no worries. 

I'll include the failures for reference:

<details>

```python
scipy/optimize/tests/test_cobyqa.py F...F..F                                               [100%]

============================================ FAILURES ============================================
________________________________ TestCOBYQA.test_minimize_simple _________________________________
scipy/optimize/tests/test_cobyqa.py:80: in test_minimize_simple
    assert_equal(sol.x, sol_new.x)
        Callback   = <class 'scipy.optimize.tests.test_cobyqa.TestCOBYQA.test_minimize_simple.<locals>.Callback'>
        CallbackNewSyntax = <class 'scipy.optimize.tests.test_cobyqa.TestCOBYQA.test_minimize_simple.<locals>.CallbackNewSyntax'>
        callback   = <scipy.optimize.tests.test_cobyqa.TestCOBYQA.test_minimize_simple.<locals>.Callback object at 0x4c9f40e0190>
        callback_new_syntax = <scipy.optimize.tests.test_cobyqa.TestCOBYQA.test_minimize_simple.<locals>.CallbackNewSyntax object at 0x4c9f40e0310>
        constraints = <scipy.optimize._constraints.NonlinearConstraint object at 0x4c9f40f0150>
        self       = <scipy.optimize.tests.test_cobyqa.TestCOBYQA object at 0x4c9e1aec810>
        sol        =  message: The lower bound for the trust-region radius has been reached
 success: True
  status: 0
     fun: 24.85185186080888
       x: [ 4.955e+00  6.667e-01]
     nit: 42
   maxcv: 8.95702001457721e-09
    nfev: 30
        sol_new    =  message: The lower bound for the trust-region radius has been reached
 success: True
  status: 0
     fun: 24.851851860808875
       x: [ 4.955e+00  6.667e-01]
     nit: 42
   maxcv: 8.957012909149853e-09
    nfev: 30
        solution   = [np.float64(4.955356249106169), 0.6666666666666666]
        x0         = [4.95, 0.66]
../../../../../.pixi/envs/free-threading/lib/python3.13t/site-packages/numpy/_utils/__init__.py:85: in wrapper
    return fun(*args, **kwargs)
        args       = (array([4.95535626, 0.66666658]), array([4.95535626, 0.66666658]), '', True)
        dep_version = '2.0.0'
        fun        = <function assert_array_equal at 0x4c9e2ffc8a0>
        kwargs     = {'strict': False}
        new_name   = 'desired'
        new_names  = ['actual', 'desired']
        old_name   = 'y'
        old_names  = ['x', 'y']
../../../../../.pixi/envs/free-threading/lib/python3.13t/contextlib.py:85: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Arrays are not equal
E   
E   Mismatched elements: 2 / 2 (100%)
E   Max absolute difference among violations: 6.28386232e-14
E   Max relative difference among violations: 9.42579469e-14
E    ACTUAL: array([4.955356, 0.666667])
E    DESIRED: array([4.955356, 0.666667])
        args       = (<built-in function eq>, array([4.95535626, 0.66666658]), array([4.95535626, 0.66666658]))
        func       = <function assert_array_compare at 0x4c9e2ffc6c0>
        kwds       = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}
        self       = <contextlib._GeneratorContextManager object at 0x4c9e07e64b0>
_________________________________ TestCOBYQA.test_minimize_array _________________________________
scipy/optimize/tests/test_cobyqa.py:187: in test_minimize_array
    assert_equal(sol.x, sol_array.x)
        bounds     = Bounds(array([4.5, 0.6]), array([5. , 0.7]))
        constraints = <scipy.optimize._constraints.NonlinearConstraint object at 0x4c9f21800b0>
        dim        = 0
        fun_array  = <function TestCOBYQA.test_minimize_array.<locals>.fun_array at 0x4c9f2060500>
        self       = <scipy.optimize.tests.test_cobyqa.TestCOBYQA object at 0x4c9e1aecf90>
        sol        =  message: The lower bound for the trust-region radius has been reached
 success: True
  status: 0
     fun: 24.851851852351608
       x: [ 4.955e+00  6.667e-01]
     nit: 36
   maxcv: 4.997531277695089e-10
    nfev: 25
        sol_array  =  message: The lower bound for the trust-region radius has been reached
 success: True
  status: 0
     fun: 24.851851852351484
       x: [ 4.955e+00  6.667e-01]
     nit: 36
   maxcv: 4.996287827907508e-10
    nfev: 25
../../../../../.pixi/envs/free-threading/lib/python3.13t/site-packages/numpy/_utils/__init__.py:85: in wrapper
    return fun(*args, **kwargs)
        args       = (array([4.95535624, 0.66666673]), array([4.95535624, 0.66666672]), '', True)
        dep_version = '2.0.0'
        fun        = <function assert_array_equal at 0x4c9e2ffc8a0>
        kwargs     = {'strict': False}
        new_name   = 'desired'
        new_names  = ['actual', 'desired']
        old_name   = 'y'
        old_names  = ['x', 'y']
../../../../../.pixi/envs/free-threading/lib/python3.13t/contextlib.py:85: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Arrays are not equal
E   
E   Mismatched elements: 2 / 2 (100%)
E   Max absolute difference among violations: 2.81543933e-09
E   Max relative difference among violations: 4.22315862e-09
E    ACTUAL: array([4.955356, 0.666667])
E    DESIRED: array([4.955356, 0.666667])
        args       = (<built-in function eq>, array([4.95535624, 0.66666673]), array([4.95535624, 0.66666672]))
        func       = <function assert_array_compare at 0x4c9e2ffc6c0>
        kwds       = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}
        self       = <contextlib._GeneratorContextManager object at 0x4c9e07e64b0>
_______________________________ TestCOBYQA.test_minimize_f_target ________________________________
scipy/optimize/tests/test_cobyqa.py:251: in test_minimize_f_target
    assert sol.nfev <= sol_ref.nfev, sol
E   AssertionError:  message: The target objective function value has been reached
E      success: True
E       status: 1
E          fun: 24.851851851938974
E            x: [ 4.955e+00  6.667e-01]
E          nit: 37
E        maxcv: 8.712319754522468e-11
E         nfev: 32
E   assert 32 <= 30
E    +  where 32 =  message: The target objective function value has been reached\n success: True\n  status: 1\n     fun: 24.851851851938974\n       x: [ 4.955e+00  6.667e-01]\n     nit: 37\n   maxcv: 8.712319754522468e-11\n    nfev: 32.nfev
E    +  and   30 =  message: The lower bound for the trust-region radius has been reached\n success: True\n  status: 0\n     fun: 24.851851860808875\n       x: [ 4.955e+00  6.667e-01]\n     nit: 42\n   maxcv: 8.957016461863532e-09\n    nfev: 30.nfev
        constraints = <scipy.optimize._constraints.NonlinearConstraint object at 0x4c9f21803d0>
        options    = {'f_target': np.float64(24.851851860808875), 'maxfev': 100}
        self       = <scipy.optimize.tests.test_cobyqa.TestCOBYQA object at 0x4c9e0628c10>
        sol        =  message: The target objective function value has been reached
 success: True
  status: 1
     fun: 24.851851851938974
       x: [ 4.955e+00  6.667e-01]
     nit: 37
   maxcv: 8.712319754522468e-11
    nfev: 32
        sol_ref    =  message: The lower bound for the trust-region radius has been reached
 success: True
  status: 0
     fun: 24.851851860808875
       x: [ 4.955e+00  6.667e-01]
     nit: 42
   maxcv: 8.957016461863532e-09
    nfev: 30
==================================== short test summary info =====================================
FAILED scipy/optimize/tests/test_cobyqa.py::TestCOBYQA::test_minimize_simple - AssertionError: 
FAILED scipy/optimize/tests/test_cobyqa.py::TestCOBYQA::test_minimize_array - AssertionError: 
FAILED scipy/optimize/tests/test_cobyqa.py::TestCOBYQA::test_minimize_f_target - AssertionError:  message: The target objective function value has been reached
```

</details>

The first two failures still give solutions, just with slightly lower precision (that may be because we're running the exact same problem in parallel though - if there's cross-talk between threads, running different problems in parallel may yield larger errors). The last failure looks a little more serious.